src/allocation_engine/interval_tree.rs

```diff
@@ -33,14 +37,139 @@ pub struct Constraint {
     pub policy: AllocPolicy,
 }
 
-/// A closed interval range [min, max] used to describe a
+/// A closed interval range [range_start, range_end] used to describe a
 /// memory slot that will be assigned to a device by the VMM.
 /// This structure represents the key of the Node object in
 /// the interval tree implementation.
-#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord, Debug)]
 pub struct Range {
-    pub min: u64,
-    pub max: u64,
+    /// Lower boundary of the interval.
+    range_start: u64,
+    /// Upper boundary of the interval.
+    range_end: u64,
+}
+
+impl Range {
+    /// Create a new Range object.
+    pub fn new(range_start: u64, range_end: u64) -> Result<Self> {
+        if range_start > range_end {
+            return Err(Error::InvalidRange(range_start, range_end));
+        }
+        Ok(Range {
+            range_start,
+            range_end,
+        })
+    }
+
+    /// Get length of the range.
+    pub fn length(&self) -> u64 {
+        self.range_end - self.range_start + 1
+    }
+
+    /// Check whether two Range objects overlaps with each other.
+    pub fn overlaps(&self, other: &Range) -> bool {
+        max(self.range_start, other.range_start) <= min(self.range_end, other.range_end)
+    }
+
+    /// Check whether the key is fully covered.
+    pub fn contains(&self, other: &Range) -> bool {
+        self.range_start <= other.range_start && self.range_end >= other.range_end
+    }
+
+    /// Create a new Range object with range_start being the first value from
+    /// the range that is aligned to `align`.
+    ///
+    /// # Examples
+    /// ```rust
+    /// extern crate vm_allocator;
+    /// use vm_allocator::allocation_engine::Range;
+    /// use vm_allocator::Result;
+    ///
+    /// fn intervals_align_forward() -> Result<Range> {
+    ///     let range_to_align = Range::new(2, 6)?;
+    ///     let res = range_to_align.align_forward(2)?;
+    ///     assert_eq!(res.length(), 5);
+    ///     Ok(res)
+    /// }
+    ///
+    /// intervals_align_forward().unwrap();
+    /// ```
+    pub fn align_forward(&self, align: u64) -> Result<Range> {
+        match align {
+            0 | 1 => Ok(*self),
+            _ => {
+                // Check alignment to be a power of two.
+                if !align.is_power_of_two() {
+                    return Err(Error::InvalidAlignment);
+                }
+                // !(align - 1)) will set all bits starting with the one already
+                // set in the alignment value, all bits until that one will be 0.
```

Comment:
I don't understand this phrase. Maybe instead of describing what `align - 1` does in text, we can just have a variable called `align_mask`. What would be important to not here is that this property only holds true for values that are a power of 2.